single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -229,12 +258,100 @@ def __init__(self, charm: AbstractMongoCharm):
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.refresh:
+            return
+
+        if self.state.app_peer_data.feature_compatibility_version == FEATURE_VERSION:
+            # We have already run all this logic before, no need to run it again.
+            return
+
+        # Update the version across all relations so that we can notify other units
+        self.cross_app_version_checker.set_version_across_all_relations()
+
+        if (
+            self.state.is_role(MongoDBRoles.CONFIG_SERVER)
+            and not self.cross_app_version_checker.are_related_apps_valid()
+        ):
+            # Early return if not all apps are valid.
+            return
+
+        try:
+            self.upgrade_backend.wait_for_cluster_healthy()  # type: ignore[attr-defined]
+        except RetryError:
+            logger.error(
+                "Cluster is not healthy after refresh, will retry next juju event.", exc_info=True
+            )
+            return
+
+        if not self.upgrade_backend.is_cluster_able_to_read_write():  # type: ignore[attr-defined]
+            logger.error(
+                "Cluster is not healthy after refresh, writes not propagated throughout cluster. Deferring post refresh check.",
+            )
+            return
+
+        try:
+            with MongoConnection(self.state.mongos_config) as mongos:
+                mongos.start_and_wait_for_balancer()
+        except BalancerNotEnabledError:
+            logger.error(
+                "Need more time to enable the balancer after finishing the refresh. Deferring event."
+            )
+            return
+
+        self.mongo_manager.set_feature_compatibility_version(FEATURE_VERSION)
+        self.state.app_peer_data.feature_compatibility_version = FEATURE_VERSION
+
+    def _post_refresh(self, refresh: charm_refresh.Common):  # noqa: C901
+        """Post refresh checks and actions.
+
+        Checks if unit is healthy and allow the next unit to update.
+        """
+        if not self.state.db_initialised:
+            return
+
+        if not refresh.workload_allowed_to_start:
+            return
+        logger.info("Restarting workloads")
+        # always apply the current charm revision's config -> no need to "migrate" configuration
+        # this charm revision's config is the one supported by the targeted workload version
+        self.dependent._configure_workloads()
+        self.dependent.start_charm_services()
+
+        self.state.unit_peer_data.current_revision = self.cross_app_version_checker.version
+
+        if self.dependent.name == CharmKind.MONGOD:
+            self.dependent._restart_related_services()  # type: ignore[attr-defined]
+
+        if self.dependent.mongo_manager.mongod_ready():
+            try:
+                self.upgrade_backend.wait_for_cluster_healthy()
+                refresh.next_unit_allowed_to_refresh = True
+            except RetryError:
+                return
```

Comment:
Could you explain how is this case handled? in the event where the cluster is unable to stabilize its health

Reply:
If the cluster is unable to stabilize its health, a message will be displayed so that the operator decides if they want to rollback or continue.
Until the cluster stabilizes, the upgrade will not be allowed to continue.
I should add a log here however to help the operator understanding what's going on.